File: backend/smart-campus-api/src/main/java/com/smartcampus/smart_campus_api/booking/service/BookingService.java
package com.smartcampus.smart_campus_api.booking.service;

import java.util.List;

import com.smartcampus.smart_campus_api.booking.dto.BookingResponse;
import com.smartcampus.smart_campus_api.booking.dto.CreateBookingRequest;

public interface BookingService {
    List<BookingResponse> getBookings(Object principal);
    BookingResponse createBooking(CreateBookingRequest request, Object principal);
    BookingResponse approveBooking(String bookingId, Object principal);
    BookingResponse rejectBooking(String bookingId, Object principal);
    BookingResponse getBookingById(String bookingId, Object principal);
    BookingResponse updateBooking(String bookingId, CreateBookingRequest request, Object principal);
    void deleteBooking(String bookingId, Object principal);
    boolean hasBookingConflict(String resourceId, String startTime, String endTime, String excludeBookingId);
}

File: frontend/smart-campus-ui/src/services/bookingService.js
// Booking service for API communication
import api from './api'

export const bookingService = {
  // Fetch all bookings from API
  async getBookings() {
    const response = await api.get('/api/bookings')
    return response.data
  },

  // Create new booking via API
  async createBooking(payload) {
    const response = await api.post('/api/bookings', payload)
    return response.data
  },

  // Approve booking via API
  async approveBooking(id) {
    const response = await api.patch(`/api/bookings/${id}/approve`)
    return response.data
  },

  // Reject booking via API
  async rejectBooking(id) {
    const response = await api.patch(`/api/bookings/${id}/reject`)
    return response.data
  },

  // Check for booking conflicts before creating/updating
  async checkBookingConflict(resourceId, startTime, endTime, excludeBookingId = null) {
    const params = new URLSearchParams({
      resourceId,
      startTime,
      endTime,
    })
    if (excludeBookingId) {
      params.append('excludeBookingId', excludeBookingId)
    }
    const response = await api.get(`/api/bookings/conflict-check?${params.toString()}`)
    return response.data
  },
}
